### PR TITLE
HW4 + HW1-4 unit tests 

### PR DESCRIPTION
**Pull request by Kuzukaze**
HW4 + HW1-4 unit tests

### CODE REVIEW

File: HW2/HW2/BitIns.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                        {
+                            mask += 1 << k;
+                        }
+                    }
+                }
+                return (a & rightPart) | b << i | (bsize > (j - i + 1) ? (a << Math.Abs(bsize - (j - i + 1)) & leftPart << bsize) : (a >> Math.Abs(bsize - (j - i + 1)) & leftPart << bsize));
+            }
+            else
+            {
+                Console.WriteLine("j must be must be greater than i and not negative");
```
Comment (NTcvilikh) [inline HW2/HW2/BitIns.cs]: Note: it's more common to check, log and return at the beginning of methods.
F.e. if (i < 0 || j < 0 || i > j) 
{ 
log.Error(...); 
return -1; 
}

File: HW2/HW2/BitIns.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                    else
+                    {
+                        leftPart += 1 << k;
+                        if (k <= j)
+                        {
+                            mask += 1 << k;
+                        }
+                    }
+                }
+                return (a & rightPart) | b << i | (bsize > (j - i + 1) ? (a << Math.Abs(bsize - (j - i + 1)) & leftPart << bsize) : (a >> Math.Abs(bsize - (j - i + 1)) & leftPart << bsize));
```
Comment (NTcvilikh) [inline HW2/HW2/BitIns.cs]: It would be better to store the result in a var before returning it 
to make further debugging easier

File: HW2/HW2/BitIns.cs (hunk 1 above)
Comment (Kuzukaze) [reply HW2/HW2/BitIns.cs]: Got it, done.

File: HW2/HW2/BitIns.cs (hunk 2 above)
Comment (Kuzukaze) [reply HW2/HW2/BitIns.cs]: done.

File: HW7/HW7/HW7_4/Program.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                }
+            }
+            return resultList;
+        }
+
+        public static string StringReverse(string str)
+        {
+            if (string.IsNullOrEmpty(str))
+            {
+                throw new ArgumentNullException("String null or empty");
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_4/Program.cs]: если можно определить поведение по умолчанию для некоректных данных,
то лучше использовать его, а не кидать исключение.
здесь, например, можно просто возвращать string.Empty.

проверки на некорректные данные и исключения нужны для того, чтобы понятно сообщить программистам-пользователям твоего метода, что они обязаны передать другие данные.

проверки на тривиальные случаи (например, пустая строка, пустой массив) нужны для того чтобы чуть ускорить работу программы (выход из метода сразу, до работы по алгоритму) и чтобы не падать (когда алгоритму требуется, например, больше элементов в массиве, чем 0 или 1, но его можно не запускать, возвращая устраивающий всех результат)

File: HW7/HW7/HW7_4/Program.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace HW7_4
+{
+    public class Program
+    {
+        static void Main()
+        {
+            var res = StringReverse("The greatest victory is that which requires no battle");
+
+            foreach (var elem in res)
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_4/Program.cs]: Console.Write(res); is enought

remove the whole foreach...

File: HW7/HW7/HW7_6.Tests/BigSumTests.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace HW7_6.Tests
+{
+    [TestFixture]
+    class BigSumTests
+    {
+        [TestCase("50000000000", "50000000000", "100000000000")]
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_6.Tests/BigSumTests.cs]: tests are not only to check that code works fine, they can also explain to other developers what is the purpose of the method, what it's doing
so... it's a good idea to add short simple test what is easy to read -- for the people
f.e. 
[TestCase("123", "1234", "1357")]

also corner-case test should be added
[TestCase("0", "0", "0")]
[TestCase("-1", "1", "0")] -- this also tells to the people that method is ready to work with negative values

File: HW7/HW7/HW7_6.Tests/BigSumTests.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        public void BigSumTest(string str1, string str2, string expectedResult)
+        {
+            var result = Program.BigSum(str1, str2);
+            Assert.AreEqual(expectedResult, result);
+        }
+
+        [TestCase(null, null)]
+        public void BigSumTeststring (string str1, string str2)
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_6.Tests/BigSumTests.cs]: BigSumTestStringIsNullOrEmpty

File: HW7/HW7/HW7_6.Tests/BigSumTests.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    class BigSumTests
+    {
+        [TestCase("50000000000", "50000000000", "100000000000")]
+        public void BigSumTest(string str1, string str2, string expectedResult)
+        {
+            var result = Program.BigSum(str1, str2);
+            Assert.AreEqual(expectedResult, result);
+        }
+
+        [TestCase(null, null)]
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_6.Tests/BigSumTests.cs]: [TestCase("", "")]
can be added - for the people and for covering corner case

File: HW7/HW7/HW7_6/Program.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            var result = BigSum("50000000000", "50000000000");
+            Console.WriteLine(result);
+        }
+
+        public static string BigSum(string str1, string str2)
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_6/Program.cs]: pls avoid static classes & static methods (unless they are really necessary)
use new Program().BigSum() instead...

File: HW7/HW7/HW7_6/Program.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            var parsed2 = BigInteger.TryParse(str2, out b);
+
+            if (parsed1 && parsed2)
+            {
+                var result = (a + b).ToString();
+                return result;
+            }
+            else
+            {
+                throw new Exception("Cant parse");
```
Comment (NTcvilikh) [inline HW7/HW7/HW7_6/Program.cs]: throw first if data is incorrect.
if (!parsed1 || !parsed2)
    throw...
then continue working... without else...

File: HW6/HW6/HW6_3/Polynomial.cs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        public double[] kArr;
+
+        public Polynomial(params double[] kArr)
+        {
+            if (kArr.Length <= 1)
+            {
+                throw new ArgumentException("Need more coefficients for Polynomial");
+            }
+            if (kArr[0] == 0)
```
Comment (NTcvilikh) [inline HW6/HW6/HW6_3/Polynomial.cs]: it's wrong to check double value equality via ==
use comparison within some epsilon instead

File: HW6/HW6/HW6_3/Polynomial.cs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace HW6_3
+{
+    public class Polynomial
+    {
+        public double[] kArr;
+
+        public Polynomial(params double[] kArr)
+        {
+            if (kArr.Length <= 1)
```
Comment (NTcvilikh) [inline HW6/HW6/HW6_3/Polynomial.cs]: it's usually possible to work with polynomials consisting of constants only, no need to throw here

File: .gitignore [hunk 12]
```diff
@@ -1,5 +1,10 @@
 ## Ignore Visual Studio temporary files, build results, and
```
Comment (NTcvilikh) [inline .gitignore]: stub comment
